mrp_bom_tracking/models/mrp_bom.py

```diff
@@ -0,0 +1,96 @@
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
+
+from odoo import fields, models
+
+
+class MrpBom(models.Model):
+    _inherit = "mrp.bom"
+
+    code = fields.Char(tracking=True)
+    product_id = fields.Many2one(tracking=True)
+    product_tmpl_id = fields.Many2one(tracking=True)
+    product_qty = fields.Float(tracking=True)
+    picking_type_id = fields.Many2one(tracking=True)
+    type = fields.Selection(tracking=True)
+
+    def write(self, values):
+        template_env = self.env["ir.ui.view"]
+        bom_line_ids = {}
+        if "bom_line_ids" in values:
+            for bom in self:
+                del_lines = []
+                for line in values["bom_line_ids"]:
+                    if line[0] == 2:
+                        del_lines.append(line[1])
+                if del_lines:
+                    message_body = template_env._render_template(
+                        "mrp_bom_tracking.track_bom_template",
+                        values={
+                            "lines": self.env["mrp.bom.line"].browse(del_lines),
+                            "mode": "Removed",
+                        },
+                    )
+                    bom.message_post(
+                        body=message_body, subtype_id=self.env.ref("mail.mt_note").id
+                    )
+                bom_line_ids[bom.id] = bom.bom_line_ids
+        res = super().write(values)
+        if "bom_line_ids" in values:
+            for bom in self:
+                new_lines = bom.bom_line_ids - bom_line_ids[bom.id]
+                if new_lines:
+                    message_body = template_env._render_template(
+                        "mrp_bom_tracking.track_bom_template",
+                        values={"lines": new_lines, "mode": "New"},
+                    )
+                    bom.message_post(
+                        body=message_body, subtype_id=self.env.ref("mail.mt_note").id
+                    )
+        return res
+
+
+class MrpBomLine(models.Model):
+    _inherit = "mrp.bom.line"
+
+    def write(self, values):
+        template_env = self.env["ir.ui.view"]
+        if "product_id" in values:
+            for bom in self.mapped("bom_id"):
+                # Bind bom variable for the current iteration to the lambda
+                lines = self.filtered(lambda x, bom=bom: x.bom_id == bom)
+                product_id = values.get("product_id")
+                if product_id:
+                    product_id = self.env["product.product"].browse(product_id)
+                product_id = product_id or lines.product_id
+                if lines:
+                    message_body = template_env._render_template(
+                        "mrp_bom_tracking.track_bom_template_2",
+                        values={"lines": lines, "product_id": product_id},
+                    )
+                    bom.message_post(
+                        body=message_body, subtype_id=self.env.ref("mail.mt_note").id
+                    )
```

Comment:
```suggestion
        if "product_id" in values:
            for bom_line in self:
                bom = bom_line.bom_id
                product_id = values.get("product_id")
                product = self.env["product.product"].browse(product_id)
                if product.exists():
                    message_body = template_env._render_template(
                        "mrp_bom_tracking.track_bom_template_2",
                        values={"lines": bom_line, "product_id": product},
                    )
                    bom.message_post(
                        body=message_body, subtype_id=self.env.ref("mail.mt_note").id
                    )
```
`product_id` check was already done at the beginning of the indentation, no need to do it again. Also, refactor to iterate over self, which contains the BoM lines, as before the change, it was already doing one message posting per line change. I believe this refactoring simplifies the understanding of the logic. What do you think?
I believe this is more of a general improvement, nothing related to the migration itself.